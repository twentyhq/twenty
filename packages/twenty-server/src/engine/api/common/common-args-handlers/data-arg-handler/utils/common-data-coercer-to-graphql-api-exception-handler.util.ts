import { assertUnreachable } from 'twenty-shared/utils';

import {
  type CommonDataCoercerException,
  CommonDataCoercerExceptionCode,
} from 'src/engine/api/common/common-args-handlers/data-arg-handler/errors/common-data-coercer.exception';
import { UserInputError } from 'src/engine/core-modules/graphql/utils/graphql-errors.util';

export const commonDataCoercerToGraphqlApiExceptionHandler = (
  error: CommonDataCoercerException,
) => {
  switch (error.code) {
    case CommonDataCoercerExceptionCode.INVALID_COMPOSITE_FIELD:
    case CommonDataCoercerExceptionCode.INVALID_NUMBER:
    case CommonDataCoercerExceptionCode.INVALID_TEXT:
    case CommonDataCoercerExceptionCode.INVALID_DATE_OR_DATE_TIME:
    case CommonDataCoercerExceptionCode.INVALID_BOOLEAN:
    case CommonDataCoercerExceptionCode.INVALID_SELECT:
    case CommonDataCoercerExceptionCode.INVALID_MULTI_SELECT:
    case CommonDataCoercerExceptionCode.INVALID_UUID:
    case CommonDataCoercerExceptionCode.INVALID_ARRAY:
    case CommonDataCoercerExceptionCode.INVALID_RAW_JSON:
      throw new UserInputError(error);
    default: {
      return assertUnreachable(error.code);
    }
  }
};
