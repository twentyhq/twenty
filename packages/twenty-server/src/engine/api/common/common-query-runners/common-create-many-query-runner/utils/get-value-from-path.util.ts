import { type ObjectRecord } from 'src/engine/api/graphql/workspace-query-builder/interfaces/object-record.interface';

export const getValueFromPath = (
  record: Partial<ObjectRecord>,
  path: string,
): string | undefined => {
  const pathParts = path.split('.');

  if (pathParts.length === 1) {
    return record[path];
  }

  const [parentField, childField] = pathParts;

  return record[parentField]?.[childField];
};
