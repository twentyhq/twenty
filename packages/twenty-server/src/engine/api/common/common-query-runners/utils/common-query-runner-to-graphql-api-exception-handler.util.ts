import {
  CommonQueryRunnerExceptionCode,
  type CommonQueryRunnerException,
} from 'src/engine/api/common/common-query-runners/errors/common-query-runner.exception';
import { NotFoundError } from 'src/engine/core-modules/graphql/utils/graphql-errors.util';

export const commonQueryRunnerToGraphqlApiExceptionHandler = (
  error: CommonQueryRunnerException,
) => {
  switch (error.code) {
    case CommonQueryRunnerExceptionCode.RECORD_NOT_FOUND:
      throw new NotFoundError(error);
    default:
      throw error;
  }
};
