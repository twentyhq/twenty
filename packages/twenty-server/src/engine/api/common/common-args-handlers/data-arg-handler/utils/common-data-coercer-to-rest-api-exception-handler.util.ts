import { BadRequestException } from '@nestjs/common';

import { assertUnreachable } from 'twenty-shared/utils';

import {
  type CommonDataCoercerException,
  CommonDataCoercerExceptionCode,
} from 'src/engine/api/common/common-args-handlers/data-arg-handler/errors/common-data-coercer.exception';

export const commonDataCoercerToRestApiExceptionHandler = (
  error: CommonDataCoercerException,
) => {
  switch (error.code) {
    case CommonDataCoercerExceptionCode.INVALID_COMPOSITE_FIELD:
    case CommonDataCoercerExceptionCode.INVALID_NUMBER:
    case CommonDataCoercerExceptionCode.INVALID_TEXT:
    case CommonDataCoercerExceptionCode.INVALID_DATE_OR_DATE_TIME:
    case CommonDataCoercerExceptionCode.INVALID_BOOLEAN:
    case CommonDataCoercerExceptionCode.INVALID_SELECT:
    case CommonDataCoercerExceptionCode.INVALID_MULTI_SELECT:
    case CommonDataCoercerExceptionCode.INVALID_UUID:
    case CommonDataCoercerExceptionCode.INVALID_ARRAY:
    case CommonDataCoercerExceptionCode.INVALID_RAW_JSON:
      throw new BadRequestException(error.message);
    default: {
      return assertUnreachable(error.code);
    }
  }
};
