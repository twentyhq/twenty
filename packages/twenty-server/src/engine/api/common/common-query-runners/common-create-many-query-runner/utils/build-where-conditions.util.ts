import { isDefined } from 'twenty-shared/utils';
import { type FindOperator, In } from 'typeorm';

import { type ObjectRecord } from 'src/engine/api/graphql/workspace-query-builder/interfaces/object-record.interface';

import { getValueFromPath } from 'src/engine/api/common/common-query-runners/common-create-many-query-runner/utils/get-value-from-path.util';

export const buildWhereConditions = (
  records: Partial<ObjectRecord>[],
  conflictingFields: {
    baseField: string;
    fullPath: string;
    column: string;
  }[],
): Record<string, FindOperator<string>>[] => {
  const whereConditions = [];

  for (const field of conflictingFields) {
    const fieldValues = records
      .map((record) => getValueFromPath(record, field.fullPath))
      .filter(isDefined);

    if (fieldValues.length > 0) {
      whereConditions.push({ [field.column]: In(fieldValues) });
    }
  }

  return whereConditions;
};
