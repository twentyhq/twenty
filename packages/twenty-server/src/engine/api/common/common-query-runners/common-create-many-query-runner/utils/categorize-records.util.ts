import { isDefined } from 'twenty-shared/utils';

import { type ObjectRecord } from 'twenty-shared/types';

import { type PartialObjectRecordWithId } from 'src/engine/api/common/common-query-runners/common-create-many-query-runner/types/partial-object-record-with-id.type';
import { getMatchingRecordId } from 'src/engine/api/common/common-query-runners/common-create-many-query-runner/utils/get-matching-record-id.util';

export const categorizeRecords = (
  records: Partial<ObjectRecord>[],
  conflictingFields: {
    baseField: string;
    fullPath: string;
    column: string;
  }[],
  existingRecords: PartialObjectRecordWithId[],
): {
  recordsToUpdate: PartialObjectRecordWithId[];
  recordsToInsert: Partial<ObjectRecord>[];
} => {
  const recordsToUpdate: PartialObjectRecordWithId[] = [];
  const recordsToInsert: Partial<ObjectRecord>[] = [];

  for (const record of records) {
    const matchingRecordId = getMatchingRecordId(
      record,
      conflictingFields,
      existingRecords,
    );

    if (isDefined(matchingRecordId)) {
      recordsToUpdate.push({ ...record, id: matchingRecordId });
    } else {
      recordsToInsert.push(record);
    }
  }

  return { recordsToUpdate, recordsToInsert };
};
