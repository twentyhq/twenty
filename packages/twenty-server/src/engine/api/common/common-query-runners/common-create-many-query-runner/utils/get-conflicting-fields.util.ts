import { capitalize } from 'twenty-shared/utils';

import { compositeTypeDefinitions } from 'src/engine/metadata-modules/field-metadata/composite-types';
import { type ObjectMetadataItemWithFieldMaps } from 'src/engine/metadata-modules/types/object-metadata-item-with-field-maps';

export const getConflictingFields = (
  objectMetadataItemWithFieldMaps: ObjectMetadataItemWithFieldMaps,
): {
  baseField: string;
  fullPath: string;
  column: string;
}[] => {
  return Object.values(objectMetadataItemWithFieldMaps.fieldsById)
    .filter((field) => field.isUnique || field.name === 'id')
    .flatMap((field) => {
      const compositeType = compositeTypeDefinitions.get(field.type);

      if (!compositeType) {
        return [
          {
            baseField: field.name,
            fullPath: field.name,
            column: field.name,
          },
        ];
      }

      const property = compositeType.properties.find(
        (prop) => prop.isIncludedInUniqueConstraint,
      );

      return property
        ? [
            {
              baseField: field.name,
              fullPath: `${field.name}.${property.name}`,
              column: `${field.name}${capitalize(property.name)}`,
            },
          ]
        : [];
    });
};
