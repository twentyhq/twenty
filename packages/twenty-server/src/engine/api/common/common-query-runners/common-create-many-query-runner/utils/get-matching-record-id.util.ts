import { msg } from '@lingui/core/macro';
import { type ObjectRecord } from 'twenty-shared/types';
import { isDefined } from 'twenty-shared/utils';

import { type PartialObjectRecordWithId } from 'src/engine/api/common/common-query-runners/common-create-many-query-runner/types/partial-object-record-with-id.type';
import { getValueFromPath } from 'src/engine/api/common/common-query-runners/common-create-many-query-runner/utils/get-value-from-path.util';
import {
  CommonQueryRunnerException,
  CommonQueryRunnerExceptionCode,
} from 'src/engine/api/common/common-query-runners/errors/common-query-runner.exception';

export const getMatchingRecordId = (
  record: Partial<ObjectRecord>,
  conflictingFields: {
    baseField: string;
    fullPath: string;
    column: string;
  }[],
  existingRecords: PartialObjectRecordWithId[],
): string | undefined => {
  const matchingRecordIds = conflictingFields.reduce<string[]>((acc, field) => {
    const requestFieldValue = getValueFromPath(record, field.fullPath);

    const matchingRecord = existingRecords.find((existingRecord) => {
      const existingFieldValue = getValueFromPath(
        existingRecord,
        field.fullPath,
      );

      return (
        isDefined(existingFieldValue) &&
        existingFieldValue === requestFieldValue
      );
    });

    if (isDefined(matchingRecord)) {
      acc.push(matchingRecord.id);
    }

    return acc;
  }, []);

  if ([...new Set(matchingRecordIds)].length > 1) {
    const conflictingFieldsValues = conflictingFields
      .map((field) => {
        const value = getValueFromPath(record, field.fullPath);

        return isDefined(value) ? `${field.fullPath}: ${value}` : undefined;
      })
      .filter(isDefined)
      .join(', ');

    throw new CommonQueryRunnerException(
      `Multiple records found with the same unique field values for ${conflictingFieldsValues}. Cannot determine which record to update.`,
      CommonQueryRunnerExceptionCode.UPSERT_MULTIPLE_MATCHING_RECORDS_CONFLICT,
      {
        userFriendlyMessage: msg`Multiple records found with the same unique field values for ${conflictingFieldsValues}. Cannot determine which record to update.`,
      },
    );
  }

  return matchingRecordIds[0];
};
