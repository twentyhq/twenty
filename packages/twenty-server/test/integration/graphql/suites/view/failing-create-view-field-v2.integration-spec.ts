import { createOneCoreViewField } from 'test/integration/metadata/suites/view-field/utils/create-one-core-view-field.util';
import { extractRecordIdsAndDatesAsExpectAny } from 'test/utils/extract-record-ids-and-dates-as-expect-any';
import {
  eachTestingContextFilter,
  type EachTestingContext,
} from 'twenty-shared/testing';
import { faker } from '@faker-js/faker';

import { type CreateViewFieldInput } from 'src/engine/core-modules/view/dtos/inputs/create-view-field.input';

import {
  cleanupViewFieldTestV2,
  setupViewFieldTestV2,
  type ViewFieldTestSetup,
} from './utils/setup-view-field-test-v2.util';

describe('View Field Resolver - Failing Create Operation - v2', () => {
  let testSetup: ViewFieldTestSetup;

  beforeAll(async () => {
    testSetup = await setupViewFieldTestV2();
  });

  afterAll(async () => {
    await cleanupViewFieldTestV2(testSetup.testObjectMetadataId);
  });

  type CreateViewFieldTestCase = (testSetup: ViewFieldTestSetup) => {
    input: CreateViewFieldInput;
  };

  const createViewFieldTestCases: EachTestingContext<CreateViewFieldTestCase>[] =
    [
      {
        title: 'non-existent field metadata',
        context: (testSetup) => ({
          input: {
            viewId: testSetup.testViewId,
            fieldMetadataId: faker.string.uuid(),
          },
        }),
      },
      {
        title: 'non-existent view metadata',
        context: (testSetup) => ({
          input: {
            viewId: faker.string.uuid(),
            fieldMetadataId: testSetup.testFieldMetadataId,
          },
        }),
      },
    ];

  it.each(eachTestingContextFilter(createViewFieldTestCases))(
    'should fail to create view field when $title',
    async ({ context }) => {
      const { input } = context(testSetup);
      const response = await createOneCoreViewField({
        input,
        expectToFail: true,
      });

      expect(response.errors).toBeDefined();
      expect(response.errors.length).toBe(1);
      const [firstError] = response.errors;

      expect(firstError).toMatchSnapshot(
        extractRecordIdsAndDatesAsExpectAny(firstError),
      );
    },
  );
});
