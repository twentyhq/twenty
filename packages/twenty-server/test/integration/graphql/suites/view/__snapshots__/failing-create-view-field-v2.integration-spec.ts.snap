// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Field Resolver - Failing Create Operation - v2 should fail to create view field when non-existent field metadata 1`] = `
{
  "extensions": {
    "code": "BAD_USER_INPUT",
    "errors": {
      "index": [],
      "objectMetadata": [],
      "view": [],
      "viewField": [
        {
          "errors": [
            {
              "code": "INVALID_VIEW_DATA",
              "message": "Field metadata not found",
              "userFriendlyMessage": "Field metadata not found",
            },
          ],
          "flatEntityMinimalInformation": {
            "fieldMetadataId": Any<String>,
            "id": Any<String>,
            "viewId": Any<String>,
          },
          "status": "fail",
          "type": "create_view_field",
        },
      ],
    },
    "message": "Validation failed for 0 object(s) and 0 field(s)",
    "summary": {
      "invalidIndex": 0,
      "invalidObjectMetadata": 0,
      "invalidView": 0,
      "invalidViewField": 0,
      "totalErrors": 0,
    },
    "userFriendlyMessage": "Validation failed for 0 object(s) and 0 field(s)",
  },
  "message": "Multiple validation errors occurred while creating view field",
  "name": "GraphQLError",
}
`;

exports[`View Field Resolver - Failing Create Operation - v2 should fail to create view field when non-existent view metadata 1`] = `
{
  "extensions": {
    "code": "BAD_USER_INPUT",
    "errors": {
      "index": [],
      "objectMetadata": [],
      "view": [],
      "viewField": [
        {
          "errors": [
            {
              "code": "INVALID_VIEW_DATA",
              "message": "View not found",
              "userFriendlyMessage": "View not found",
            },
          ],
          "flatEntityMinimalInformation": {
            "fieldMetadataId": Any<String>,
            "id": Any<String>,
            "viewId": Any<String>,
          },
          "status": "fail",
          "type": "create_view_field",
        },
      ],
    },
    "message": "Validation failed for 0 object(s) and 0 field(s)",
    "summary": {
      "invalidIndex": 0,
      "invalidObjectMetadata": 0,
      "invalidView": 0,
      "invalidViewField": 0,
      "totalErrors": 0,
    },
    "userFriendlyMessage": "Validation failed for 0 object(s) and 0 field(s)",
  },
  "message": "Multiple validation errors occurred while creating view field",
  "name": "GraphQLError",
}
`;
